### PR TITLE
#1170 expand service invocation coverage

### PR DESCRIPTION
**Pull request by LMWF**
# Description

Adds more cases for dapr service invocation:

1. http client -> http server
2. grpc client -> http server
3. http client -> grpc server
4. grpc client -> grpc server

The first 2 categories also try with the 4 currently supported http methods (verbs).  Also adds validation of data in body both ways.  All methods send a json-ized string and return a custom object.

Adds another app that receives gRPC for categories 2 and 4.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #1170

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation
* [ ] Specification has been updated
* [ ] Provided sample for the feature

### CODE REVIEW

**Comment by LMWF** [comment]
@youngbupark did the linter not run on the current 2 commits?  It seems it it ran with only 1 commit.

File: tests/apps/service_invocation_grpc/app.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+// This method gets invoked when a remote service has called the app through Dapr
+// The payload carries a Method to identify the method, a set of metadata properties and an optional payload
+func (s *server) OnInvoke(ctx context.Context, in *pb.InvokeEnvelope) (*any.Any, error) {
+	fmt.Printf("Got invoked method %s and data: %s\n", in.Method, string(in.Data.Value))
+
+	var err error
+	var response []byte
+	switch in.Method {
+	case "httpToGrpcTest":
```
Comment (artursouza) [inline tests/apps/service_invocation_grpc/app.go]: Use "if" statement or https://github.com/golang/go/wiki/Switch#fall-through

**Comment by youngbupark** [comment]
> @youngbupark did the linter not run on the current 2 commits? It seems it it ran with only 1 commit. Lines it's pointing to don't exist anymore.

The history includes only one commit.

File: tests/apps/service_invocation/service.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      labels:
+        testapp: service-invocation
+      annotations:
+        dapr.io/enabled: "true"
+        dapr.io/id: "myapp"
+        dapr.io/port: "3000"
+    spec:
+      containers:
+      - name: service-invocation
+        image: docker.io/[YOUR ALIAS]/e2e-service_invocation:dev
```
Comment (artursouza) [inline tests/apps/service_invocation/service.yaml]: Is this one for http and the other one for grpc? If so, I would expect this one to contain "http" just like the other image contains "grpc".

**Comment by LMWF** [comment]
It shows 2 for me in the "commits" tab.

**Comment by youngbupark** [comment]
> It shows 2 for me in the "commits" tab.

Ah with the original commit. I re-ran the build. see how it goes.

File: tests/e2e/service_invocation/service_invocation_test.go [hunk 3]
```diff
@@ -89,27 +97,91 @@ var serviceinvocationTests = []struct {
...
 		t.Run(tt.in, func(t *testing.T) {
 			body, err := json.Marshal(testCommandRequest{
 				RemoteApp: tt.remoteApp,
-				Method:    tt.method,
+				Method:    tt.appMethod,
+			})
+			require.NoError(t, err)
+
+			resp, err := utils.HTTPPost(
+				fmt.Sprintf("%s/tests/invoke_test", externalURL), body)
```
Comment (artursouza) [inline tests/e2e/service_invocation/service_invocation_test.go]: Why you don't need the appMethod here?

File: tests/e2e/service_invocation/service_invocation_test.go [hunk 4]
```diff
@@ -89,27 +97,91 @@ var serviceinvocationTests = []struct {
...
+			body, err := json.Marshal(testCommandRequest{
+				RemoteApp: tt.remoteApp,
+				Method:    tt.appMethod,
 			})
 			require.NoError(t, err)
 
-			resp, err := utils.HTTPPost(fmt.Sprintf("%s/tests/invoke_test", externalURL), body)
+			url := fmt.Sprintf("http://%s/%s", externalURL, tt.appMethod)
+
+			t.Logf("url is '%s'\n", url)
```
Comment (artursouza) [inline tests/e2e/service_invocation/service_invocation_test.go]: This code block is almost identical to the one above. Any reason they cannot be combined?

File: tests/e2e/service_invocation/service_invocation_test.go (hunk 3 above)
Comment (LMWF) [reply tests/e2e/service_invocation/service_invocation_test.go]: appMethod is the method to call via dapr service invokcation.  /tests/invoke_tests, is a non-dapr endpoint used just to talk to the app.  That handler body which contains appMethod then calls appMethod via dapr from there.

File: tests/e2e/service_invocation/service_invocation_test.go (hunk 4 above)
Comment (LMWF) [reply tests/e2e/service_invocation/service_invocation_test.go]: The url has a different format here, and usage of appMethod is different.  In this case unlike above it really is the name of the method for the test to call (again not thru service invocation though).  I thought it was easier and more readable to keep them separate.

File: tests/apps/service_invocation/service.yaml (hunk 2 above)
Comment (LMWF) [reply tests/apps/service_invocation/service.yaml]: I just used names matching the existing app which I added to.  This app now has "both", though for grpc it only has a client, it doesn't listen on grpc.

**Pull request merged**